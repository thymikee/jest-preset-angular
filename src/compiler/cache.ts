import type * as ts from 'typescript';

/**
 * Copy from https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/ivy/cache.ts
 */
export class SourceFileCache extends Map<string, ts.SourceFile> {
  readonly #angularDiagnostics = new Map<ts.SourceFile, ts.Diagnostic[]>();

  invalidate(file: string): void {
    const sourceFile = this.get(file);
    if (sourceFile) {
      this.delete(file);
      this.#angularDiagnostics.delete(sourceFile);
    }
  }

  updateAngularDiagnostics(sourceFile: ts.SourceFile, diagnostics: ts.Diagnostic[]): void {
    if (diagnostics.length > 0) {
      this.#angularDiagnostics.set(sourceFile, diagnostics);
    } else {
      this.#angularDiagnostics.delete(sourceFile);
    }
  }

  getAngularDiagnostics(sourceFile: ts.SourceFile): ts.Diagnostic[] | undefined {
    return this.#angularDiagnostics.get(sourceFile);
  }
}
