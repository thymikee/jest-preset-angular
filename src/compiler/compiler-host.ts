import type { Logger } from 'bs-logger';
import { normalize } from 'path';
import type { TTypeScript } from 'ts-jest/dist/types';
import { LINE_FEED } from 'ts-jest/dist/constants';
import type * as ts from 'typescript';

import type { NgJestConfig } from '../config/ng-jest-config';
import type { CompilerOptions } from '@angular/compiler-cli';

export class NgJestCompilerHost implements ts.CompilerHost {
  private readonly _sourceFileCache: Map<string, ts.SourceFile> = new Map<string, ts.SourceFile>();
  private readonly _memoryHost: Map<string, string | undefined> = new Map<string, string | undefined>();
  private readonly _emittedResult: [string, string] = ['', ''];
  private readonly _ts: TTypeScript;
  private readonly _moduleResolutionCache: ts.ModuleResolutionCache;

  constructor(private logger: Logger, private ngJestCfg: NgJestConfig) {
    this._ts = this.ngJestCfg.compilerModule;
    this._moduleResolutionCache = this._ts.createModuleResolutionCache(this.ngJestCfg.cwd, (x) => x);
  }

  getEmittedResult(): [string, string] {
    return this._emittedResult;
  }

  updateMemoryHost(fileName: string, fileContent: string): void {
    const previousContents = this._memoryHost.get(fileName);
    const contentsChanged = previousContents !== fileContent;
    if (contentsChanged) {
      this._memoryHost.set(fileName, fileContent);
    }
  }

  getSourceFile(fileName: string, languageVersion: ts.ScriptTarget): ts.SourceFile | undefined {
    const normalizedFileName = normalize(fileName);
    try {
      const cached = this._sourceFileCache.get(normalizedFileName);
      if (cached) {
        return cached;
      }

      this.logger.debug(
        { normalizedFileName },
        'getSourceFile: cache miss, reading file and create source file to update cache',
      );

      const content = this.readFile(normalizedFileName);
      if (content !== undefined) {
        const sf = this._ts.createSourceFile(normalizedFileName, content, languageVersion, true);
        this._sourceFileCache.set(fileName, sf);

        return sf;
      }
    } catch (e) {
      this.logger.error(e);
    }

    return undefined;
  }

  getDefaultLibFileName(options: ts.CompilerOptions): string {
    return this._ts.createCompilerHost(options).getDefaultLibFileName(options);
  }

  // This is due to typescript CompilerHost interface being weird on writeFile. This shuts down
  // typings in WebStorm.
  get writeFile() {
    return (fileName: string, data: string, _writeByteOrderMark: boolean): void => {
      this._emittedResult[fileName.endsWith('.map') ? 1 : 0] = data;
    };
  }

  getCurrentDirectory(): string {
    return this.ngJestCfg.cwd;
  }

  getCanonicalFileName(fileName: string): string {
    return this.useCaseSensitiveFileNames() ? fileName : fileName.toLowerCase();
  }

  useCaseSensitiveFileNames(): boolean {
    return this._ts.sys.useCaseSensitiveFileNames;
  }

  getNewLine(): string {
    return LINE_FEED;
  }

  fileExists(fileName: string): boolean {
    return this._ts.sys.fileExists(fileName);
  }

  readFile(fileName: string): string | undefined {
    const normalizedFileName = normalize(fileName);
    let fileContent = this._memoryHost.get(normalizedFileName);
    if (!fileContent) {
      this.logger.debug(
        { fileName: normalizedFileName },
        'readFile: file does not exist in memory cache, read file with file system',
      );

      fileContent = this._ts.sys.readFile(normalizedFileName) ?? undefined;
      this._memoryHost.set(normalizedFileName, fileContent);
    }

    return fileContent;
  }

  trace(message: string): void {
    this.logger.trace(message);
  }

  resolveModuleNames(
    moduleNames: string[],
    containingFile: string,
    _reusedNames: string[] | undefined,
    redirectedReference: ts.ResolvedProjectReference | undefined,
    options: CompilerOptions,
  ): (ts.ResolvedModule | undefined)[] {
    return moduleNames.map((moduleName) => {
      const { resolvedModule } = this._ts.resolveModuleName(
        moduleName,
        containingFile,
        options,
        this._ts.createCompilerHost(options),
        this._moduleResolutionCache,
        redirectedReference,
      );

      return resolvedModule;
    });
  }

  resolveTypeReferenceDirectives(
    typeReferenceDirectiveNames: string[],
    containingFile: string,
    redirectedReference: ts.ResolvedProjectReference | undefined,
    options: CompilerOptions,
  ): (ts.ResolvedTypeReferenceDirective | undefined)[] {
    return typeReferenceDirectiveNames.map((typeDirectiveName) => {
      const { resolvedTypeReferenceDirective } = this._ts.resolveTypeReferenceDirective(
        typeDirectiveName,
        containingFile,
        options,
        this._ts.createCompilerHost(options),
        redirectedReference,
      );

      return resolvedTypeReferenceDirective;
    });
  }
}
