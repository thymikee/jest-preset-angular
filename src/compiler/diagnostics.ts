import type { Diagnostic, Diagnostics } from '@angular/compiler-cli';
import { CancellationToken, DiagnosticMode, hasErrors } from '@ngtools/webpack/src/diagnostics';
import type * as ts from 'typescript';

/**
 * Copy from `@ngtools/webpack` and strip out stuffs
 */
export function gatherDiagnostics(
  program: ts.Program,
  mode = DiagnosticMode.All,
  cancellationToken?: CancellationToken,
): Diagnostics {
  const allDiagnostics: Array<ts.Diagnostic | Diagnostic> = [];
  let checkOtherDiagnostics = true;

  // eslint-disable-next-line @typescript-eslint/ban-types
  function checkDiagnostics<T extends Function>(fn: T) {
    if (checkOtherDiagnostics) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const diags = fn(undefined, cancellationToken);
      if (diags) {
        allDiagnostics.push(...diags);

        checkOtherDiagnostics = !hasErrors(diags);
      }
    }
  }

  const gatherSyntacticDiagnostics = (mode & DiagnosticMode.Syntactic) != 0;
  const gatherSemanticDiagnostics = (mode & DiagnosticMode.Semantic) != 0;
  const tsProgram = program;
  if (gatherSyntacticDiagnostics) {
    // Check syntactic diagnostics.
    checkDiagnostics(tsProgram.getSyntacticDiagnostics.bind(tsProgram));
  }
  if (gatherSemanticDiagnostics) {
    checkDiagnostics(tsProgram.getSemanticDiagnostics.bind(tsProgram));
  }

  return allDiagnostics;
}
