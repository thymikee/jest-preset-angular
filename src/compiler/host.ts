import ts from 'typescript';

import { normalizePath } from './paths';

export const augmentHostWithCaching = (host: ts.CompilerHost, cache: Map<string, ts.SourceFile>): void => {
  const baseGetSourceFile = host.getSourceFile;
  host.getSourceFile = function (fileName, languageVersion, onError, shouldCreateNewSourceFile, ...parameters) {
    if (!shouldCreateNewSourceFile && cache.has(fileName)) {
      return cache.get(fileName);
    }

    const file = baseGetSourceFile.call(host, fileName, languageVersion, onError, true, ...parameters);

    if (file) {
      cache.set(fileName, file);
    }

    return file;
  };
};

/**
 * Augments a TypeScript Compiler Host's resolveModuleNames function to collect dependencies
 * of the containing file passed to the resolveModuleNames function. This process assumes
 * that consumers of the Compiler Host will only call resolveModuleNames with modules that are
 * actually present in a containing file.
 * This process is a workaround for gathering a TypeScript SourceFile's dependencies as there
 * is no currently exposed public method to do so. A BuilderProgram does have a `getAllDependencies`
 * function.
 */
export const augmentHostWithDependencyCollection = (
  host: ts.CompilerHost,
  dependencies: Map<string, Set<string>>,
  moduleResolutionCache?: ts.ModuleResolutionCache,
): void => {
  if (host.resolveModuleNames) {
    const baseResolveModuleNames = host.resolveModuleNames;
    host.resolveModuleNames = function (moduleNames: string[], containingFile: string, ...parameters) {
      const results = baseResolveModuleNames.call(host, moduleNames, containingFile, ...parameters);

      const containingFilePath = normalizePath(containingFile);
      for (const result of results) {
        if (result) {
          const containingFileDependencies = dependencies.get(containingFilePath);
          if (containingFileDependencies) {
            containingFileDependencies.add(result.resolvedFileName);
          } else {
            dependencies.set(containingFilePath, new Set([result.resolvedFileName]));
          }
        }
      }

      return results;
    };
  } else {
    host.resolveModuleNames = function (
      moduleNames: string[],
      containingFile: string,
      _reusedNames: string[] | undefined,
      redirectedReference: ts.ResolvedProjectReference | undefined,
      options: ts.CompilerOptions,
    ) {
      return moduleNames.map((name) => {
        const result = ts.resolveModuleName(
          name,
          containingFile,
          options,
          host,
          moduleResolutionCache,
          redirectedReference,
        ).resolvedModule;

        if (result) {
          const containingFilePath = normalizePath(containingFile);
          const containingFileDependencies = dependencies.get(containingFilePath);
          if (containingFileDependencies) {
            containingFileDependencies.add(result.resolvedFileName);
          } else {
            dependencies.set(containingFilePath, new Set([result.resolvedFileName]));
          }
        }

        return result;
      });
    };
  }
};
