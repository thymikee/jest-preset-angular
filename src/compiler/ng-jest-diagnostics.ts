import ts from 'typescript';

const ERROR_CODE_MATCHER = /(\u001b\[\d+m ?)TS-99(\d+: ?\u001b\[\d+m)/g;

/**
 * During formatting of `ts.Diagnostic`s, the numeric code of each diagnostic is prefixed with the
 * hard-coded "TS" prefix. For Angular's own error codes, a prefix of "NG" is desirable. To achieve
 * this, all Angular error codes start with "-99" so that the sequence "TS-99" can be assumed to
 * correspond with an Angular specific error code. This function replaces those occurrences with
 * just "NG".
 *
 * @param errors The formatted diagnostics
 */
const replaceTsWithNgInErrors = (errors: string): string => {
  return errors.replace(ERROR_CODE_MATCHER, '$1NG$2');
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isTsDiagnostic = (diagnostic: any): diagnostic is ts.Diagnostic => {
  return !!diagnostic && diagnostic.source !== 'angular';
};

const defaultFormatHost: ts.FormatDiagnosticsHost = {
  getCurrentDirectory: () => ts.sys.getCurrentDirectory(),
  getCanonicalFileName: (fileName) => fileName,
  getNewLine: () => ts.sys.newLine,
};

type Diagnostics = readonly ts.Diagnostic[];

export const formatDiagnostics = (diags: Diagnostics, host: ts.FormatDiagnosticsHost = defaultFormatHost): string => {
  if (diags.length) {
    return diags
      .map((diagnostic) => {
        if (isTsDiagnostic(diagnostic)) {
          return replaceTsWithNgInErrors(ts.formatDiagnosticsWithColorAndContext([diagnostic], host));
        } else {
          return ts.formatDiagnostics(diagnostic, host);
        }
      })
      .join('');
  } else {
    return '';
  }
};

interface NgJestCompilation {
  warnings: Error[];
  errors: Error[];
}

export type DiagnosticsReporter = (diagnostics: Diagnostics) => void;

export const createDiagnosticsReporter = (
  compilation: NgJestCompilation,
  formatter: (diagnostic: Diagnostics[number]) => string,
): DiagnosticsReporter => {
  return (diagnostics) => {
    for (const diagnostic of diagnostics) {
      const text = formatter(diagnostic);
      if (diagnostic.category === ts.DiagnosticCategory.Error) {
        addError(compilation, text);
      } else {
        addWarning(compilation, text);
      }
    }
  };
};

const addWarning = (compilation: NgJestCompilation, message: string): void => {
  compilation.warnings.push(new Error(message));
};

const addError = (compilation: NgJestCompilation, message: string): void => {
  compilation.errors.push(new Error(message));
};
