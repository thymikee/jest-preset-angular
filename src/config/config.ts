import type { TsJestTransformerOptions } from 'ts-jest';

export type NgJestTransformerOptions = TsJestTransformerOptions & {
    processWithEsbuild?: string[];
};
