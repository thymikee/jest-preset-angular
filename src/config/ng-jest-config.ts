import type { Logger } from 'bs-logger';
import { globsToMatcher } from 'jest-util';
import { type ProjectConfigTsJest, type RawCompilerOptions, ConfigSet } from 'ts-jest';
import type { ParsedCommandLine } from 'typescript';

/**
 * Some NPM packages like `tslib` is distributed in such a way that `esbuild` cannot process it, so we fall back to use TypeScript API
 */
const defaultProcessWithEsbuildPatterns = ['**/*.mjs'];

export class NgJestConfig extends ConfigSet {
  readonly processWithEsbuild: ReturnType<typeof globsToMatcher>;

  constructor(jestConfig: ProjectConfigTsJest | undefined, parentLogger?: Logger | undefined) {
    super(jestConfig, parentLogger);
    const jestGlobalsConfig = jestConfig?.globals?.ngJest ?? Object.create(null);
    this.processWithEsbuild = globsToMatcher([
      ...(jestGlobalsConfig.processWithEsbuild ?? []),
      ...defaultProcessWithEsbuildPatterns,
    ]);
  }

  /**
   * Override `ts-jest` behavior because we use `readConfiguration` which will read and resolve tsconfig.
   */
  protected _resolveTsConfig(compilerOptions?: RawCompilerOptions, resolvedConfigFile?: string): ParsedCommandLine {
    const result = super._resolveTsConfig(compilerOptions, resolvedConfigFile) as ParsedCommandLine;
    result.options.enableIvy = true;
    result.options.noEmitOnError = false;
    result.options.suppressOutputPathCheck = true;
    result.options.allowEmptyCodegenFiles = false;
    result.options.annotationsAs = 'decorators';
    result.options.enableResourceInlining = false;
    // Since we define preset default also transform `js` so we need to set `allowJs` true
    result.options.allowJs = true;

    return result;
  }
}
