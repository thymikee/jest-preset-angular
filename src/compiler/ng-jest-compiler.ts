import { CompilerHost, CompilerOptions, createCompilerHost } from '@angular/compiler-cli';
import type { Logger } from 'bs-logger';
import { updateOutput } from 'ts-jest/dist/compiler/compiler-utils';
import type { CompilerInstance, TTypeScript, ResolvedModulesMap } from 'ts-jest/dist/types';
import type * as ts from 'typescript';

import type { NgJestConfig } from '../config/ng-jest-config';
import { factory as downlevelCtor } from '../transformers/downlevel-ctor';
import { factory as inlineFiles } from '../transformers/inline-files';
import { factory as stripStyles } from '../transformers/strip-styles';
import { NgJestCompilerHost } from './compiler-host';

export class NgJestCompiler implements CompilerInstance {
  private _compilerOptions!: CompilerOptions;
  private _program: ts.Program | undefined;
  private _compilerHost: CompilerHost | undefined;
  private _tsHost: NgJestCompilerHost | undefined;
  private _rootNames: string[] = [];
  private readonly _logger: Logger;
  private readonly _ts: TTypeScript;

  constructor(readonly ngJestConfig: NgJestConfig, readonly jestCacheFS: Map<string, string>) {
    this._logger = this.ngJestConfig.logger;
    this._ts = this.ngJestConfig.compilerModule;
    this._setupOptions(this.ngJestConfig);

    this._logger.debug('created NgJestCompiler');
  }

  getResolvedModulesMap(fileContent: string, fileName: string): ResolvedModulesMap {
    this._tsHost?.updateMemoryHost(fileName, fileContent);

    // eslint-disable-next-line
    return (this._program?.getSourceFile(fileName) as any)?.resolvedModules;
  }

  getCompiledOutput(fileName: string, fileContent: string, supportsStaticESM: boolean): string {
    const customTransformers = this.ngJestConfig.customTransformers;
    const transformers = {
      ...customTransformers,
      before: [
        // hoisting from `ts-jest` or other before transformers
        ...(customTransformers.before as ts.TransformerFactory<ts.SourceFile>[]),
        inlineFiles(this.ngJestConfig),
        stripStyles(this.ngJestConfig),
      ],
    };
    if (this._program) {
      const allDiagnostics = [];
      if (!this._rootNames.includes(fileName)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: update memory host, rootFiles and Program');

        this._rootNames = [...this._rootNames, fileName];
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        this._tsHost!.updateMemoryHost(fileName, fileContent);
        this._createOrUpdateProgram();
      }

      this._logger.debug({ fileName }, 'getCompiledOutput: compiling using Program');

      const sourceFile = this._program.getSourceFile(fileName);
      const emitResult = this._program.emit(sourceFile, undefined, undefined, undefined, {
        ...transformers,
        before: [
          // hoisting from `ts-jest` or other before transformers
          ...transformers.before,
          /**
           * Downlevel constructor parameters for DI support. This is required to support forwardRef in ES2015 due to
           * TDZ issues. This wrapper is needed here due to the program not being available until after
           * the transformers are created. Also because program can be updated so we can't push this transformer in
           * _createCompilerHost
           */
          downlevelCtor(this._program),
        ],
      });
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const compiledOutput: [string, string] = this._tsHost!.getEmittedResult();
      if (this.ngJestConfig.shouldReportDiagnostics(fileName)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: getting diagnostics');

        allDiagnostics.push(
          ...emitResult.diagnostics,
          ...this._program.getSyntacticDiagnostics(sourceFile),
          ...this._program.getSemanticDiagnostics(sourceFile),
        );
      }
      if (!allDiagnostics.length) {
        this._logger.debug({ fileName }, 'getCompiledOutput: update compiled output including inline source map');

        return updateOutput(compiledOutput[0], fileName, compiledOutput[1]);
      } else {
        this.ngJestConfig.raiseDiagnostics(allDiagnostics, fileName);

        return '';
      }
    } else {
      let moduleKind = this._compilerOptions.module;
      if (supportsStaticESM && this.ngJestConfig.useESM) {
        moduleKind =
          !moduleKind ||
          (moduleKind &&
            ![this._ts.ModuleKind.ES2015, this._ts.ModuleKind.ES2020, this._ts.ModuleKind.ESNext].includes(moduleKind))
            ? this._ts.ModuleKind.ESNext
            : moduleKind;
      } else {
        moduleKind = this._ts.ModuleKind.CommonJS;
      }

      this._logger.debug({ fileName }, 'getCompiledOutput: compiling as isolated module');

      const result: ts.TranspileOutput = this._ts.transpileModule(fileContent, {
        fileName,
        transformers,
        compilerOptions: {
          ...this._compilerOptions,
          module: moduleKind,
        },
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        reportDiagnostics: this.ngJestConfig.shouldReportDiagnostics(fileName),
      });
      if (result.diagnostics && this.ngJestConfig.shouldReportDiagnostics(fileName)) {
        this.ngJestConfig.raiseDiagnostics(result.diagnostics, fileName, this._logger);
      }

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      return updateOutput(result.outputText, fileName, result.sourceMapText!);
    }
  }

  private _setupOptions({ parsedTsConfig }: NgJestConfig) {
    this._logger.debug({ parsedTsConfig }, '_setupOptions: initializing compiler config');

    this._compilerOptions = { ...parsedTsConfig.options };
    this._rootNames = parsedTsConfig.rootNames.filter((rootName) => !this.ngJestConfig.isTestFile(rootName));
    if (this._compilerOptions.strictMetadataEmit) {
      this._logger.warn(
        `Using Angular compiler option 'strictMetadataEmit' for applications might cause undefined behavior.`,
      );
    }
    if (!this.ngJestConfig.isolatedModules) {
      this._logger.debug(
        { compilerOptions: this._compilerOptions },
        '_setupOptions: creating Compiler Host using @angular/compiler-cli createCompilerHost',
      );

      this._tsHost = new NgJestCompilerHost(this._logger, this.ngJestConfig, this.jestCacheFS);
      this._compilerHost = createCompilerHost({
        options: this._compilerOptions,
        tsHost: this._tsHost,
      });
      this._createOrUpdateProgram();
    }
  }

  private _createOrUpdateProgram(): void {
    const oldTsProgram = this._program;

    this._logger.debug(`_createOrUpdateProgram: ${oldTsProgram ? 'update' : 'create'} TypeScript Program`);

    this._program = this._ts.createProgram(this._rootNames, this._compilerOptions, this._compilerHost, oldTsProgram);
  }
}
