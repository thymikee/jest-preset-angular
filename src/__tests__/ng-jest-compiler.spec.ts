import { readFileSync } from 'fs';
import { join } from 'path';

import { jest } from '@jest/globals';
import ts from 'typescript';

import { NgJestCompiler } from '../compiler/ng-jest-compiler';
import { NgJestConfig } from '../config/ng-jest-config';

import { jestCfgStub } from './__helpers__/test-constants';
import { mockFolder } from './__helpers__/test-helpers';

import SpyInstance = jest.SpyInstance;

describe('NgJestCompiler', () => {
  describe('with isolatedModules true', () => {
    let transpileModuleSpy: SpyInstance;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const baseJestCfg = {
      ...jestCfgStub,
      globals: {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        'ts-jest': {
          ...jestCfgStub.globals['ts-jest'],
          isolatedModules: true,
        },
      },
    };

    beforeEach(() => {
      // @ts-expect-error testing purpose
      transpileModuleSpy = ts.transpileModule = jest.fn().mockReturnValueOnce({
        outputText: 'var foo = 1',
        diagnostics: [],
        sourceMapText: '{}',
      });
    });

    test('should call transpileModule with CommonJS module', () => {
      const ngJestConfig = new NgJestConfig(baseJestCfg);
      const fileName = join(mockFolder, 'foo.service.ts');
      const compiler = new NgJestCompiler(ngJestConfig, new Map());

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'), false)!;

      expect(transpileModuleSpy).toHaveBeenCalled();
      expect(transpileModuleSpy.mock.calls[0][1].compilerOptions.module).toEqual(ts.ModuleKind.CommonJS);
    });

    test('should call transpileModule with ESM module', () => {
      const ngJestConfig = new NgJestConfig({
        ...baseJestCfg,
        globals: {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
          'ts-jest': {
            ...baseJestCfg.globals['ts-jest'],
            useESM: true,
          },
        },
      });
      const fileName = join(mockFolder, 'foo.service.ts');
      const compiler = new NgJestCompiler(ngJestConfig, new Map());

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'), true)!;

      expect(transpileModuleSpy).toHaveBeenCalled();
      expect(transpileModuleSpy.mock.calls[0][1].compilerOptions.module).not.toEqual(ts.ModuleKind.CommonJS);
    });
  });

  describe('with isolatedModule false', () => {
    const sourceMapStub =
      '{"version":3,"file":"app.component.js","sourceRoot":"","sources":["app.component.ts"],"names":[],"mappings":";;;;AAAA,wCAA0C;IAO7B,YAAY,SAAZ,YAAY;;QACvB,UAAK,GAAG,cAAc,CAAC;IACzB,CAAC;CAAA,CAAA;AAFY,YAAY;IALxB,gBAAS,CAAC;QACT,QAAQ,EAAE,UAAU;QACpB,WAAW,EAAE,sBAAsB;QACnC,SAAS,EAAE,CAAC,sBAAsB,CAAC;KACpC,CAAC;GACW,YAAY,CAExB;AAFY,oCAAY","sourcesContent":["import { Component } from \'@angular/core\';\\n\\n@Component({\\n  selector: \'app-root\',\\n  templateUrl: \'./app.component.html\',\\n  styleUrls: [\'./app.component.scss\'],\\n})\\nexport class AppComponent {\\n  title = \'test-app-v10\';\\n}\\n"]}';
    const compiledOutputStub =
      '"use strict";\n' +
      'Object.defineProperty(exports, "__esModule", { value: true });\n' +
      'exports.AppComponent = void 0;\n' +
      'const tslib_1 = require("tslib");\n' +
      'const core_1 = require("@angular/core");\n' +
      'let AppComponent = class AppComponent {\n' +
      '    constructor() {\n' +
      "        this.title = 'test-app-v10';\n" +
      '    }\n' +
      '};\n' +
      'AppComponent = tslib_1.__decorate([\n' +
      '    core_1.Component({\n' +
      "        selector: 'app-root',\n" +
      "        templateUrl: './app.component.html',\n" +
      "        styleUrls: ['./app.component.scss'],\n" +
      '    })\n' +
      '], AppComponent);\n' +
      'exports.AppComponent = AppComponent;\n' +
      '//# sourceMappingURL=app.component.js.map\n';

    const ngJestConfig = new NgJestConfig(jestCfgStub);
    const noErrorFileName = join(mockFolder, 'app.component.ts');
    const noErrorFileContent = readFileSync(noErrorFileName, 'utf-8');
    const hasErrorFileName = join(mockFolder, 'foo.component.ts');
    const hasErrorFileContent = readFileSync(hasErrorFileName, 'utf-8');

    test.each([noErrorFileName, undefined])(
      'should return compiled result for new file which is not known or known by Program',
      (fileName) => {
        ngJestConfig.parsedTsConfig = {
          ...ngJestConfig.parsedTsConfig,
          rootNames: fileName ? [fileName] : [],
        };
        const compiler = new NgJestCompiler(ngJestConfig, new Map());
        // @ts-expect-error bypass type checking to access private property
        compiler._tsHost.getEmittedResult = jest.fn().mockReturnValueOnce([compiledOutputStub, sourceMapStub]);

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        compiler.getCompiledOutput(noErrorFileName, noErrorFileContent, false)!;

        // @ts-expect-error bypass type checking to access private property
        // eslint-disable-next-line @typescript-eslint/unbound-method
        expect(compiler._tsHost.getEmittedResult).toHaveBeenCalled();
      },
    );

    test.each([hasErrorFileName, undefined])('should throw diagnostics error for new file which is', (fileName) => {
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: fileName ? [fileName] : [],
      };
      const compiler = new NgJestCompiler(ngJestConfig, new Map());

      expect(() =>
        compiler.getCompiledOutput(hasErrorFileName, hasErrorFileContent, false),
      ).toThrowErrorMatchingSnapshot(fileName ? 'known by Program' : 'not known by Program');
    });

    test('should not throw diagnostics error when shouldReportDiagnostics return false', () => {
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: [hasErrorFileName],
      };
      // @ts-expect-error testing purpose
      ngJestConfig.shouldReportDiagnostics = jest.fn().mockReturnValueOnce(false);
      const compiler = new NgJestCompiler(ngJestConfig, new Map());

      expect(() => compiler.getCompiledOutput(hasErrorFileName, hasErrorFileContent, false)).not.toThrowError();
    });
  });
});
