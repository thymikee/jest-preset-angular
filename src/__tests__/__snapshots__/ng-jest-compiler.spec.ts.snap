// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgJestCompiler with isolatedModule false should throw diagnostics error for new file which is: known by Program 1`] = `"src/__tests__/__mocks__/foo.component.ts(8,3): error TS2322: Type '\\"test-app-v10\\"' is not assignable to type 'number'."`;

exports[`NgJestCompiler with isolatedModule false should throw diagnostics error for new file which is: not known by Program 1`] = `"src/__tests__/__mocks__/foo.component.ts(8,3): error TS2322: Type '\\"test-app-v10\\"' is not assignable to type 'number'."`;
